trillian/ctfe/dns_handlers.go

```diff
@@ -0,0 +1,324 @@
+
+const (
+	// Used for STH responses, which change over time.
+	sthTTL = 5 * 60
+	// Used for other responses, which are immutable
+	respTTL = 24 * 60 * 60
+)
+
+var (
+	// In these formats the last (.*) is always the zone, this is checked outside
+	// the regex matching.
+	sthRE     = regexp.MustCompile(`^sth\.(.*)\.$`)
+	consistRE = regexp.MustCompile(`^(\d+)\.(\d+)\.(\d+)\.sth-consistency\.(.*)\.$`)
+	hashRE    = regexp.MustCompile(`^([A-Z0-9]+)\.hash\.(.*)\.$`)
+	treeRE    = regexp.MustCompile(`^(\d+)\.(\d+)\.(\d+)\.tree\.(.*)\.$`)
+)
+
+type dnsFunc func(context.Context, *CTDNSHandler, []string, dns.ResponseWriter, *dns.Msg)
+
+// CTDNSHandler encapsulates all the logic for serving CT related DNS
+// requests.
+type CTDNSHandler struct {
+	cfg      *configpb.LogConfig
+	logCtx   *LogContext
+	handlers []dnsHandler
+}
+
+type dnsHandler struct {
+	matchRE  *regexp.Regexp
+	handleFn dnsFunc
+}
+
+// NewDNS creates a new DNS handler for use with a particular CT log.
+func NewDNS(cfg *configpb.LogConfig, logCtx *LogContext) dns.Handler {
+	return &CTDNSHandler{
+		cfg:    cfg,
+		logCtx: logCtx,
+		handlers: []dnsHandler{
+			{matchRE: sthRE, handleFn: sthFunc},
+			{matchRE: consistRE, handleFn: consistFunc},
+			{matchRE: hashRE, handleFn: hashFunc},
+			{matchRE: treeRE, handleFn: treeFunc},
+		},
+	}
+}
+
+// ServeDNS implements the dns.Handler interface and is the main serving
+// API.
+func (c *CTDNSHandler) ServeDNS(w dns.ResponseWriter, r *dns.Msg) {
+	// From the spec all requests must be class INET, type TXT.
+	if err := validate(r); err != nil {
+		glog.Warningf("CTDNSHandler.ServeDNS(): %v", err)
+		failWithRcode(w, r, dns.RcodeFormatError, err)
+		return
+	}
+
+	glog.V(1).Infof("CTDNSHandler.ServeDNS(): Query: %v", r.Question[0].Name)
+	q := dns.Fqdn(r.Question[0].Name)
+	ctx, cancel := context.WithTimeout(context.Background(), c.logCtx.instanceOpts.Deadline)
+	defer cancel()
+
+	// Find the first handler that accepts the name pattern match.
+	for _, h := range c.handlers {
+		params := h.matchRE.FindStringSubmatch(q)
+		// All our matches should include at least the zone as a capture group
+		// so we must have at least two results from the above.
+		if len(params) > 1 {
+			// Additionally check that the zone matched the last regex param and the
+			// whole string matched to avoid any false positives.
+			if params[0] == q && params[len(params)-1] == c.cfg.DnsZone {
+				// This handler accepted the match and provides the result.
+				h.handleFn(ctx, c, params, w, r)
+				return
+			}
+		}
+	}
+	// No handler matched. Reject the request.
+	failWithRcode(w, r, dns.RcodeNotZone, nil)
+}
+
+func sthFunc(ctx context.Context, c *CTDNSHandler, params []string, w dns.ResponseWriter, r *dns.Msg) {
+	sth, err := GetTreeHead(ctx, c.logCtx.rpcClient, c.logCtx.logID, c.logCtx.LogPrefix)
+	if err != nil {
+		failWithRcode(w, r, dns.RcodeServerFailure, err)
+		return
+	}
+
+	// Add the signature over the STH contents.
```

Comment:
Hmm. This isn't going to work for a mirror since the operator presumably doesn't have the key.

Would it be better to keep this as a completely separate thing which is a DNS server on the front, and a CT client on the back (and thus works exactly the same for logs & mirrors?)

Reply:
This is what CTFE does so that's going to need changes for mirroring as well. Whatever we do there should "just work" (tm) here. Not sure about making it an HTTP client and why we didn't do that for the C++ version?